File: sdk-core/src/test/java/care/data4life/sdk/AttachmentServiceTest.java
/*
 * Copyright (c) 2020 D4L data4life gGmbH / All rights reserved.
 *
 * D4L owns all legal rights, title and interest in and to the Software Development Kit ("SDK"),
 * including any intellectual property rights that subsist in the SDK.
 *
 * The SDK and its documentation may be accessed and used for viewing/review purposes only.
 * Any usage of the SDK for other purposes, including usage for the development of
 * applications/third-party applications shall require the conclusion of a license agreement
 * between you and D4L.
 *
 * If you are interested in licensing the SDK for your own applications/third-party
 * applications and/or if you’d like to contribute to the development of the SDK, please
 * contact D4L by email to help@data4life.care.
 */

package care.data4life.sdk;

import org.junit.Before;
import org.junit.Test;

import java.util.Arrays;
import java.util.List;
import java.util.Objects;

import care.data4life.crypto.GCKey;
import care.data4life.fhir.stu3.model.Attachment;
import care.data4life.fhir.stu3.model.FhirDateTime;
import care.data4life.fhir.stu3.util.FhirDateTimeParser;
import care.data4life.sdk.config.DataRestrictionException;
import care.data4life.sdk.helpers.AttachmentBuilder;
import care.data4life.sdk.lang.D4LException;
import care.data4life.sdk.lang.DataValidationException;
import io.reactivex.Single;
import io.reactivex.observers.TestObserver;
import kotlin.Pair;

import static com.google.common.truth.Truth.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

public class AttachmentServiceTest {

    private static final String ALIAS = "alias";
    private static final String ATTACHMENT_ID = "attachmentId";
    private static final String THUMBNAIL_ID = "attachmentId#previewId";
    private static final String RECORD_ID = "recordId";
    private static final String USER_ID = "userId";
    private static final String TITLE = "title";
    private static final String CONTENT_TYPE = "contentType";
    private final FhirDateTime creationDate = FhirDateTimeParser.parseDateTime("2013-04-03");
    private final byte[] pdf = new byte[]{0x25, 0x50, 0x44, 0x46, 0x2d};
    private byte[] largeFile = new byte[20000000 + 1];
    private static final String dataBase64 = "JVBERi0=";
    private static final String DATA_HASH = "dataHash";
    private final GCKey attachmentKey = mock(GCKey.class);

    private FileService mockFileService;
    private ImageResizer mockImageResizer;
    private Attachment attachment;
    private AttachmentService attachmentService; //SUT

    @Before
    public void setUp() throws DataRestrictionException.UnsupportedFileType, DataRestrictionException.MaxDataSizeViolation {
        attachment = AttachmentBuilder.buildWith(TITLE, creationDate, CONTENT_TYPE, pdf);
        mockFileService = mock(FileService.class);
        mockImageResizer = mock(ImageResizer.class);

        attachmentService = new AttachmentService(ALIAS, mockFileService, mockImageResizer);
    }

    @Test
    public void uploadAttachments_shouldReturnListOfAttachments() throws InterruptedException, DataRestrictionException.UnsupportedFileType, DataRestrictionException.MaxDataSizeViolation {
        // given
        attachment.id = "id";
        Attachment newAttachment = AttachmentBuilder.buildWith("newAttachment", creationDate, CONTENT_TYPE, pdf);
        newAttachment.id = null;
        List<Attachment> attachments = Arrays.asList(attachment, newAttachment);

        when(mockFileService.uploadFile(attachmentKey, USER_ID, pdf)).thenReturn(Single.just(ATTACHMENT_ID));

        // when
        TestObserver<List<Pair<Attachment, List<String>>>> subscriber = attachmentService
                .uploadAttachments(attachments, attachmentKey, USER_ID).test().await();

        // then
        List<Pair<Attachment, List<String>>> result = subscriber
                .assertNoErrors()
                .assertComplete()
                .assertValueCount(1)
                .values().get(0);

        assertThat(result).hasSize(2);
        Attachment a1 = result.get(0).getFirst();
        assertThat(a1.id).isEqualTo(ATTACHMENT_ID);
        assertThat(a1.title).isEqualTo(TITLE);
        assertThat(a1.creation).isEqualTo(creationDate);
        assertThat(a1.contentType).isEqualTo(CONTENT_TYPE);
        assertThat(a1.data).isEqualTo(dataBase64);

        Attachment a2 = result.get(1).getFirst();
        assertThat(a2.id).isEqualTo(ATTACHMENT_ID);
        assertThat(a2.title).isEqualTo("newAttachment");
        assertThat(a2.creation).isEqualTo(creationDate);
        assertThat(a2.contentType).isEqualTo(CONTENT_TYPE);
        assertThat(a2.data).isEqualTo(dataBase64);

        verify(mockFileService, times(2)).uploadFile(attachmentKey, USER_ID, pdf);
        verifyNoMoreInteractions(mockFileService);
    }

    @Test
    public void uploadAttachment_with_20MB_and_1_shouldFail() throws InterruptedException, DataRestrictionException.UnsupportedFileType, DataRestrictionException.MaxDataSizeViolation {
        // given
        attachment.id = "id";
        System.arraycopy(pdf, 0, largeFile, 0, pdf.length);
        try {
            Attachment newAttachment = AttachmentBuilder.buildWith("newAttachment", creationDate, CONTENT_TYPE, largeFile);
            newAttachment.id = null;
            List<Attachment> attachments = Arrays.asList(attachment, newAttachment);
            //fail("Should have thrown an exception");
        } catch (DataRestrictionException.MaxDataSizeViolation e) {
            assert (true);
        }

    }

    @Test
    public void downloadAttachments_shouldReturnListOfAttachments() throws InterruptedException, DataValidationException.InvalidAttachmentPayloadHash {
        // given
        attachment.id = ATTACHMENT_ID;
        attachment.data = null;
        List<Attachment> attachments = Arrays.asList(attachment);
        when(mockFileService.downloadFile(attachmentKey, USER_ID, ATTACHMENT_ID)).thenReturn(Single.just(pdf));

        // when
        TestObserver<List<Attachment>> subscriber =
                attachmentService.downloadAttachments(attachments, attachmentKey, USER_ID).test().await();

        // then
        List<Attachment> result = subscriber
                .assertNoErrors()
                .assertComplete()
                .assertValueCount(1)
                .values().get(0);

        assertThat(result).hasSize(1);
        Attachment a = result.get(0);
        assertThat(a.id).isEqualTo(ATTACHMENT_ID);
        assertThat(a.title).isEqualTo(TITLE);
        assertThat(a.creation).isEqualTo(creationDate);
        assertThat(a.contentType).isEqualTo(CONTENT_TYPE);
        assertThat(a.data).isEqualTo(dataBase64);

        verify(mockFileService).downloadFile(attachmentKey, USER_ID, ATTACHMENT_ID);
        verifyNoMoreInteractions(mockFileService);
    }

    @Test
    public void downloadAttachments_shouldThrow_whenInvalidHashAttachment() throws DataValidationException.InvalidAttachmentPayloadHash {
        // Given
        attachment.id = ATTACHMENT_ID;
        attachment.hash = DATA_HASH;
        List<Attachment> attachments = Arrays.asList(attachment);

        // When
        try {
            attachmentService.downloadAttachments(attachments, attachmentKey, USER_ID);
            //fail("Exception expected!");
        } catch (D4LException e) {

            // Then
            assertThat(e.getClass()).isEqualTo(DataValidationException.InvalidAttachmentPayloadHash.class);
            assertThat(e.getMessage()).isEqualTo("Attachment.hash is not valid");
        }
    }

    @Test
    public void downloadAttachments_shouldTNot_throw_whenInvalidHashPreview() throws InterruptedException, DataValidationException.InvalidAttachmentPayloadHash {
        // Given
        attachment.id = THUMBNAIL_ID;
        attachment.hash = DATA_HASH;
        List<Attachment> attachments = Arrays.asList(attachment);
        when(mockFileService.downloadFile(attachmentKey, USER_ID, THUMBNAIL_ID.split(RecordService.SPLIT_CHAR)[1])).thenReturn(Single.just(pdf));

        // when
        TestObserver<List<Attachment>> subscriber =
                attachmentService.downloadAttachments(attachments, attachmentKey, USER_ID).test().await();

        // then
        List<Attachment> result = subscriber
                .assertNoErrors()
                .assertComplete()
                .assertValueCount(1)
                .values().get(0);

        assertThat(result).hasSize(1);
        Attachment a = result.get(0);
        assertThat(a.id).isEqualTo(THUMBNAIL_ID);
        assertThat(a.title).isEqualTo(TITLE);
        assertThat(a.creation).isEqualTo(creationDate);
        assertThat(a.contentType).isEqualTo(CONTENT_TYPE);
        assertThat(a.data).isEqualTo(dataBase64);

        verify(mockFileService).downloadFile(attachmentKey, USER_ID, THUMBNAIL_ID.split(RecordService.SPLIT_CHAR)[1]);
        verifyNoMoreInteractions(mockFileService);

    }

    @Test
    public void updatingAttachments_shouldReturnListOfAttachmentsInOrder() throws InterruptedException, DataRestrictionException.UnsupportedFileType, DataRestrictionException.MaxDataSizeViolation {
        // given
        attachment.id = null;
        Attachment newAttachment = AttachmentBuilder.buildWith("newAttachment", creationDate, CONTENT_TYPE, pdf);
        List<Attachment> attachments = Arrays.asList(attachment, newAttachment);

        when(mockFileService.uploadFile(attachmentKey, USER_ID, pdf)).thenReturn(Single.just(ATTACHMENT_ID));

        // when
        TestObserver<List<Pair<Attachment, List<String>>>> subscriber =
                attachmentService.uploadAttachments(attachments, attachmentKey, USER_ID).test().await();

        // then
        List<Pair<Attachment, List<String>>> result = subscriber
                .assertNoErrors()
                .assertComplete()
                .assertValueCount(1)
                .values().get(0);

        assertThat(result).hasSize(2);
        assertThat(result.get(0).getFirst()).isEqualTo(attachment);
        assertThat(result.get(1).getFirst()).isEqualTo(newAttachment);
    }

    @Test
    public void deleteAttachment() throws InterruptedException {
        // given
        when(mockFileService.deleteFile(USER_ID, ATTACHMENT_ID)).thenReturn(Single.just(true));

        // when
        TestObserver<Boolean> subscriber = attachmentService.deleteAttachment(ATTACHMENT_ID, USER_ID).test().await();

        // then
        subscriber.assertNoErrors()
                .assertComplete()
                .assertValue(it -> it);
    }

    @Test
    public void deleteAttachment_shouldFail() throws InterruptedException {
        // given
        when(mockFileService.deleteFile(ATTACHMENT_ID, USER_ID)).thenReturn(Single.error(new Throwable()));

        // when
        TestObserver<Boolean> subscriber = attachmentService.deleteAttachment(ATTACHMENT_ID, USER_ID).test().await();

        // then
        subscriber.assertError(Objects::nonNull)
                .assertNotComplete();
    }
}

File: sdk-core/src/test/java/care/data4life/sdk/attachment/AttachmentServiceTest.kt
/*
 * Copyright (c) 2020 D4L data4life gGmbH / All rights reserved.
 *
 * D4L owns all legal rights, title and interest in and to the Software Development Kit ("SDK"),
 * including any intellectual property rights that subsist in the SDK.
 *
 * The SDK and its documentation may be accessed and used for viewing/review purposes only.
 * Any usage of the SDK for other purposes, including usage for the development of
 * applications/third-party applications shall require the conclusion of a license agreement
 * between you and D4L.
 *
 * If you are interested in licensing the SDK for your own applications/third-party
 * applications and/or if you’d like to contribute to the development of the SDK, please
 * contact D4L by email to help@data4life.care.
 */

package care.data4life.sdk.attachment

import care.data4life.sdk.crypto.GCKey
import care.data4life.sdk.test.util.GenericTestDataProvider.ATTACHMENT_ID
import care.data4life.sdk.test.util.GenericTestDataProvider.USER_ID
import care.data4life.sdk.util.Base64.encodeToString
import care.data4life.sdk.wrapper.SdkImageResizer
import care.data4life.sdk.wrapper.WrapperContract
import io.mockk.Runs
import io.mockk.clearAllMocks
import io.mockk.every
import io.mockk.just
import io.mockk.mockk
import io.mockk.mockkConstructor
import io.mockk.mockkObject
import io.mockk.slot
import io.mockk.unmockkObject
import io.mockk.verify
import io.mockk.verifyOrder
import io.reactivex.Single
import org.junit.Before
import org.junit.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertSame
import kotlin.test.assertTrue

class AttachmentServiceTest {
    private val fileService: AttachmentContract.FileService = mockk()
    private val resizer: AttachmentContract.ImageResizer = mockk()
    private lateinit var service: AttachmentContract.Service

    @Before
    fun setUp() {
        clearAllMocks()
        mockkConstructor(SdkImageResizer::class)

        service = AttachmentService(fileService, resizer)
    }

    @Test
    fun `It fulfils AttachmentService`() {
        val service: Any = AttachmentService(mockk(), mockk())

        assertTrue(service is AttachmentContract.Service)
    }

    @Test
    fun `Given delete is called with a AttachmentId and UserId it delegates the call to the FileService and returns its result`() {
        // Given
        val attachmentId = ATTACHMENT_ID
        val userId = USER_ID

        val expected = Single.just(true)

        every { fileService.deleteFile(userId, attachmentId) } returns expected

        // When
        val actual = service.delete(attachmentId, userId)

        // Then
        assertSame(
            actual = actual,
            expected = expected
        )
    }

    @Test
    fun `Given download is called with a List of Attachments, the AttachmentKey and the UserId, filters out Attachments, which have no Id`() {
        // Given
        val attachment1: WrapperContract.Attachment = mockk()
        val attachment2: WrapperContract.Attachment = mockk()
        val attachmentKey: GCKey = mockk()

        every { attachment1.id } returns null
        every { attachment2.id } returns null

        // When
        val downloaded = service.download(
            listOf(attachment1, attachment2),
            attachmentKey,
            USER_ID
        ).blockingGet()

        // Then
        assertEquals(
            actual = downloaded.size,
            expected = 0
        )
    }

    @Test
    fun `Given download is called with a List of Attachments, the AttachmentKey and the UserId, it returns a list of Attachments with data`() {
        // Given
        mockkObject(AttachmentDownloadHelper)

        val userID = USER_ID
        val attachmentId = "id1"
        val derivedId = "id2"
        val attachment: WrapperContract.Attachment = mockk()
        val attachmentKey: GCKey = mockk()
        val downloadedAttachment = ByteArray(23)

        every { attachment.id } returns attachmentId

        every {
            fileService.downloadFile(attachmentKey, userID, derivedId)
        } returns Single.just(downloadedAttachment)

        every { AttachmentDownloadHelper.deriveAttachmentId(attachment) } returns derivedId
        every {
            AttachmentDownloadHelper.addAttachmentPayload(attachment, downloadedAttachment)
        } returns attachment

        // When
        val downloadedAttachments = service.download(
            listOf(attachment),
            attachmentKey,
            userID
        ).blockingGet()

        // Then
        assertEquals(
            expected = 1,
            actual = downloadedAttachments.size
        )
        assertSame(
            actual = downloadedAttachments[0],
            expected = attachment
        )

        verifyOrder {
            AttachmentDownloadHelper.deriveAttachmentId(attachment)
            fileService.downloadFile(attachmentKey, userID, derivedId)
            AttachmentDownloadHelper.addAttachmentPayload(attachment, downloadedAttachment)
        }

        unmockkObject(AttachmentDownloadHelper)
    }

    @Test
    fun `Given upload is called with a list of Attachments, which contain data, the AttachmentKey and the UserId, it filters attachments which have actually no data`() {
        // Given
        val attachment1: WrapperContract.Attachment = mockk()
        val attachment2: WrapperContract.Attachment = mockk()
        val attachmentKey: GCKey = mockk()

        every { attachment1.data } returns null
        every { attachment2.data } returns null

        // When
        val uploaded = service.upload(
            listOf(attachment1, attachment2),
            attachmentKey,
            USER_ID
        ).blockingGet()

        // Then
        assertEquals(
            actual = uploaded.size,
            expected = 0
        )

        verify(exactly = 0) { fileService.uploadFile(any(), any(), any()) }
    }

    @Test
    fun `Given upload is called with a list of Attachments, which contain data, the AttachmentKey and the UserId, it returns the uploaded attachments and no AdditionalIds if the Attachment is not resizeable`() {
        // Given
        val userId = USER_ID
        val receivedId = "newId"
        val attachment: WrapperContract.Attachment = mockk()
        val data = "test".toByteArray()
        val encodedData = encodeToString(data)
        val attachmentKey: GCKey = mockk()

        every { attachment.id = receivedId } just Runs

        every { attachment.data } returns encodedData
        every {
            fileService.uploadFile(attachmentKey, userId, data)
        } returns Single.just(receivedId)

        every { anyConstructed<SdkImageResizer>().isResizable(data) } returns false

        // When
        val uploaded = service.upload(listOf(attachment), attachmentKey, userId).blockingGet()

        // Then
        assertEquals(
            actual = uploaded.size,
            expected = 1
        )

        val (modifiedAttachment, additionalIds) = uploaded[0]
        assertTrue(additionalIds is List<*>)
        assertEquals(
            actual = additionalIds.size,
            expected = 0
        )
        assertSame(
            actual = modifiedAttachment,
            expected = attachment
        )

        verify(exactly = 1) { attachment.id = receivedId }
        verify(exactly = 1) { fileService.uploadFile(any(), any(), any()) }
    }

    @Test
    fun `Given upload is called with a list of Attachments, which contain data, the AttachmentKey and the UserId, it returns the uploaded attachments and null for AdditionalIds if the Attachment is resizeable, but the resizing fails`() {
        // Given
        val userId = USER_ID
        val receivedId = "newId"
        val attachment: WrapperContract.Attachment = mockk()
        val data = "test".toByteArray()
        val encodedData = encodeToString(data)
        val attachmentKey: GCKey = mockk()
        val originalData = slot<ByteArray>()

        every { attachment.id } returns receivedId
        every { attachment.id = receivedId } just Runs

        every { attachment.data } returns encodedData
        every {
            fileService.uploadFile(attachmentKey, userId, data)
        } returns Single.just(receivedId)

        every { anyConstructed<SdkImageResizer>().isResizable(data) } returns true
        every {
            anyConstructed<SdkImageResizer>().resize(capture(originalData), any())
        } answers { originalData.captured }

        // When
        val uploaded = service.upload(listOf(attachment), attachmentKey, userId).blockingGet()

        // Then
        assertEquals(
            actual = uploaded.size,
            expected = 1
        )

        val (modifiedAttachment, additionalIds) = uploaded[0]
        assertNull(additionalIds)
        assertSame(
            actual = modifiedAttachment,
            expected = attachment
        )

        verify(exactly = 1) { attachment.id = receivedId }
        verify(exactly = 1) { fileService.uploadFile(any(), any(), any()) }
    }

    @Test
    fun `Given upload is called with a list of Attachments, which contain data, the AttachmentKey and the UserId, it returns the uploaded attachments and AdditionalIds if the Attachment is resizeable and the resizing succeeds`() {
        // Given
        val userId = USER_ID
        val receivedId = "newId"
        val attachment: WrapperContract.Attachment = mockk()
        val data = "test".toByteArray()
        val encodedData = encodeToString(data)
        val attachmentKey: GCKey = mockk()

        val preview = ByteArray(23)
        val thumbnail = ByteArray(42)

        val previewId = "prev"
        val thumbnailId = "thumb"

        every { attachment.id } returns receivedId
        every { attachment.id = receivedId } just Runs

        every { attachment.data } returns encodedData
        every {
            fileService.uploadFile(attachmentKey, userId, data)
        } returns Single.just(receivedId)

        every { anyConstructed<SdkImageResizer>().isResizable(data) } returns true
        every {
            anyConstructed<SdkImageResizer>().resize(
                data,
                AttachmentContract.ImageResizer.DEFAULT_PREVIEW_SIZE_PX,
            )
        } returns preview

        every {
            fileService.uploadFile(attachmentKey, userId, preview)
        } returns Single.just(previewId)

        every {
            anyConstructed<SdkImageResizer>().resize(
                data,
                AttachmentContract.ImageResizer.DEFAULT_THUMBNAIL_SIZE_PX,
            )
        } returns thumbnail

        every {
            fileService.uploadFile(attachmentKey, userId, thumbnail)
        } returns Single.just(thumbnailId)

        // When
        val uploaded = service.upload(listOf(attachment), attachmentKey, userId).blockingGet()

        // Then
        assertEquals(
            actual = uploaded.size,
            expected = 1
        )

        val (modifiedAttachment, additionalIds) = uploaded[0]
        assertTrue(additionalIds is List<*>)
        assertEquals(
            actual = additionalIds.size,
            expected = 2
        )
        assertEquals(
            actual = additionalIds[0],
            expected = previewId
        )
        assertEquals(
            actual = additionalIds[1],
            expected = thumbnailId
        )

        assertSame(
            actual = modifiedAttachment,
            expected = attachment
        )

        verify(exactly = 1) { attachment.id = receivedId }
        verify(exactly = 3) { fileService.uploadFile(any(), any(), any()) }
    }

    @Test
    fun `Given upload is called with a list of Attachments, which contain data, the AttachmentKey and the UserId, it returns the uploaded attachments and AdditionalIds with a duplicate of the AttachmentId, if the image is too small`() {
        // Given
        val userId = USER_ID
        val receivedId = "newId"
        val attachment: WrapperContract.Attachment = mockk()
        val data = "test".toByteArray()
        val encodedData = encodeToString(data)
        val attachmentKey: GCKey = mockk()

        val preview = ByteArray(23)
        val thumbnail = ByteArray(42)

        val previewId = "prev"
        val thumbnailId = "thumb"

        every { attachment.id } returns receivedId
        every { attachment.id = receivedId } just Runs

        every { attachment.data } returns encodedData
        every {
            fileService.uploadFile(attachmentKey, userId, data)
        } returns Single.just(receivedId)

        every { anyConstructed<SdkImageResizer>().isResizable(data) } returns true
        every {
            anyConstructed<SdkImageResizer>().resize(
                data,
                AttachmentContract.ImageResizer.DEFAULT_PREVIEW_SIZE_PX,
            )
        } returns preview

        every {
            fileService.uploadFile(attachmentKey, userId, preview)
        } returns Single.just(previewId)

        every {
            anyConstructed<SdkImageResizer>().resize(
                data,
                AttachmentContract.ImageResizer.DEFAULT_THUMBNAIL_SIZE_PX,
            )
        } returns null

        every {
            fileService.uploadFile(attachmentKey, userId, thumbnail)
        } returns Single.just(thumbnailId)

        // When
        val uploaded = service.upload(listOf(attachment), attachmentKey, userId).blockingGet()

        // Then
        assertEquals(
            actual = uploaded.size,
            expected = 1
        )

        val (modifiedAttachment, additionalIds) = uploaded[0]
        assertTrue(additionalIds is List<*>)
        assertEquals(
            actual = additionalIds.size,
            expected = 2
        )
        assertEquals(
            actual = additionalIds[0],
            expected = previewId
        )
        assertEquals(
            actual = additionalIds[1],
            expected = receivedId
        )

        assertSame(
            actual = modifiedAttachment,
            expected = attachment
        )

        verify(exactly = 1) { attachment.id = receivedId }
        verify(exactly = 2) { fileService.uploadFile(any(), any(), any()) }
    }
}
